src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -2194,6 +2187,21 @@ private static async Task ExecuteResultWriteResponse(IResult? result, HttpContex
         await EnsureRequestResultNotNull(result).ExecuteAsync(httpContext);
     }
 
+    private static Task WriteJsonResponse(HttpResponse response, object? value)
+    {
+        if (value == null)
+        {
+            return HttpResponseJsonExtensions.WriteAsJsonAsync(response, value, default);
+        }
+
+        // Call WriteAsJsonAsync() with the runtime type to serialize the runtime type rather than the declared type
+        // and avoid source generators issues.
+        // https://github.com/dotnet/aspnetcore/issues/43894
+        // https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-polymorphism
+        return HttpResponseJsonExtensions.WriteAsJsonAsync(response, value, value.GetType(), default);
```

Comment:
Nit: The top version is just calling the generic version with `<object>` which I think is correct but can be simplified

```suggestion
        // Call WriteAsJsonAsync() with the runtime type to serialize the runtime type rather than the declared type
        // and avoid source generators issues.
        // https://github.com/dotnet/aspnetcore/issues/43894
        // https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-polymorphism
        return HttpResponseJsonExtensions.WriteAsJsonAsync(response, value, value?.GetType() ?? typeof(object), default);
```